stonesoup/sensormanager/reward.py

```diff
@@ -80,57 +81,43 @@ def __call__(self, config: Mapping[Sensor, Sequence[Action]], tracks: Set[Track]
         # Reward value
         config_metric = 0
 
-        # Create dictionary of predictions for the tracks in the configuration
-        predictions = {track: self.predictor.predict(track[-1],
-                                                     timestamp=metric_time)
-                       for track in tracks}
-        # Running updates
-        r_updates = dict()
-
         predicted_sensors = list()
-
         memo = {}
         # For each sensor in the configuration
         for sensor, actions in config.items():
-            predicted_sensor = deepcopy(sensor, memo)
+            predicted_sensor = copy.deepcopy(sensor, memo)
             predicted_sensor.add_actions(actions)
             predicted_sensor.act(metric_time)
             if isinstance(sensor, Sensor):
                 predicted_sensors.append(predicted_sensor)  # checks if its a sensor
 
-        for sensor in predicted_sensors:
-            # some logic needed here to check if sensor is platform...
-
-            for track in tracks:
-
-                # If the track is selected by a sensor for the first time -
-                # 'previous' is the prediction
-                # If the track has already been selected by a sensor -
-                # 'previous' is the most recent update
-                if track not in r_updates:
-                    previous = predictions[track]
-                else:
-                    previous = r_updates[track]
-
-                previous_cov_norm = np.linalg.norm(previous.covar)
-
-                predicted_track = Track(previous, init_metadata=dict(Length=3, Width=1))
+        # Create dictionary of predictions for the tracks in the configuration
+        predicted_tracks = set()
+        for track in tracks:
+            predicted_track = copy.copy(track)
+            predicted_track.states = copy.copy(predicted_track.states)
+            predicted_track.metadatas = copy.copy(predicted_track.metadatas)
```

Comment:
Why do you need to copy everything separately here and why are they copies of themselves? (Probably just me not understanding!)

Reply:
When you append states it'll modify the original list (and metadatas list), which you don't want. But if you deep copy the entire track to avoid this, it'll also copy all the states of the track which isn't required and super slow for any track with a lengthy history.

I should add a comment to make this clear in the code.

Reply:
In fact, having thought about this, I think we should implement a proper shallow copy method like a list/dictionary would have. I'll create a new PR for this once this is merged. It'll be same as here effectively, but part of the class, so a plain `copy.copy(track)` will ensure you get a shallow copy, and adding/removing states won't effect the original.